aspnetcore/fundamentals/http-requests.md

```diff
@@ -0,0 +1,394 @@
+
+An interface and a reply are defined to represent the external API and its response:
+
+```csharp
+public interface IHelloClient
+{
+    [Get("/helloworld")]
+    Task<Reply> GetMessageAsync();
+}
+
+public class Reply
+{
+    public string Message { get; set; }
+}
+```
+
+A typed client can be added, using Refit to generate the implementation:
+
+```csharp
+public void ConfigureServices(IServiceCollection services)
+{
+    services.AddHttpClient("hello", c =>
+    {
+        c.BaseAddress = new Uri("http://localhost:5000");
+    })
+    .AddTypedClient(c => Refit.RestService.For<IHelloClient>(c));
+
+    services.AddMvc();
+}
+```
+
+The defined interface can be consumed where necessary, with the implementation provided by DI and Refit:
+
+```csharp
+[ApiController]
+public class ValuesController : ControllerBase
+{
+    private readonly IHelloClient _client;
+
+    public ValuesController(IHelloClient client)
+    {
+        _client = client;
+    }
+    
+    [HttpGet("/")]
+    public async Task<ActionResult<Reply>> Index()
+    {
+        return await _client.GetMessageAsync();
+    }
+}
+```
+
+## Outgoing request middleware
+
+`HttpClient` already has the concept of delegating handlers that can be linked together for outgoing HTTP requests. The `HttpClientFactory` makes registration of per-named clients more intuitive. It supports registration and chaining of multiple handlers together to build an outgoing request middleware pipeline. Each of these handlers is able to perform work before and after the outgoing request. This pattern is similar to the inbound middleware pipeline in ASP.NET Core. This provides a mechanism to manage cross-cutting concerns around HTTP requests, including caching, error handling, serialization, and logging.
+
+To create a handler, a class can be added, deriving from `DelegatingHandler`. The `SendAsync` method can then be overridden to execute code before passing the request to the next handler in the pipeline:
+
+```csharp
+private class RetryHandler : DelegatingHandler
+{
+    public int RetryCount { get; set; } = 5;
+
+    protected override Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+    {
+        for (var i = 0; i < RetryCount; i++)
+        {
+            try
+            {
+                return base.SendAsync(request, cancellationToken);
+            }
+            catch (HttpRequestException) when (i == RetryCount - 1)
+            {
+                throw;
+            }
+            catch (HttpRequestException)
+            {
+                // Retry
+                Task.Delay(TimeSpan.FromMilliseconds(50));
+            }
+        }
+
+        // Unreachable.
+        throw null;
+    }
+}
+```
+
+The preceding code defines a basic retry handler which retries up to five times if a `HttpRequestException` is caught.
```

Comment:
I'm not sure if we have an example of a handler that isn't a dupe of what's in Polly, but our actual guidance will be to use Polly for this kind of thing.

Maybe generating and adding a tracking requestId?

Reply:
Yeah I was trying to come up with something realistic that wouldn't be solved by Polly and hit a wall. I can look at something which generates a tracking ID and puts it into the header for the sample.

Reply:
yeah, the two ideas we have here are to do service discovery or to do something like a tracking header. /cc @glennc  for pointer